src/workerd/io/worker.c++

```diff
@@ -3736,6 +3740,33 @@ void Worker::Actor::shutdown(uint16_t reasonCode, kj::Maybe<const kj::Exception&
   // capability server should have triggered this (potentially indirectly) via its destructor.
   KJ_IF_SOME(r, impl->ioContext) {
     impl->metrics->shutdown(reasonCode, r.get()->getLimitEnforcer());
+
+    // Dispatch unload event for DurableObjectState if one was attached
+    KJ_IF_SOME(handler, onUnload) {
+      worker->runInLockScope(
```

Comment:
Nit: this could be moved into a utility function to avoid duplication with `~IoContext`... but that can be done later.